# <center>Laboratorium Analiza i bazy danych </center>

## <center>Łączenie tabel, podzapytania i funkcje agregujące</center>

PIOTR SUCHY 407332

## Przykładowe tabele obrazujące łączenie

Do zobrazowania operacji łączenia zostaną użyte tabele:

```sql
CREATE TABLE shape_a (
    id INT PRIMARY KEY,
    shape VARCHAR (100) NOT NULL
);
 
CREATE TABLE shape_b (
    id INT PRIMARY KEY,
    shape VARCHAR (100) NOT NULL
);
```
 
Polecenie CREATE TABLE tworzy tabelę o zadanej nazwie i strukturze. Ogólna postać to:
```sql
CREATE TABLE tab_name (
    col_name1 data_type constrain,
    col_name1 data_type constrain,
    ...
);
```
Należy uzupełnić ją danymi:
```sql
INSERT INTO shape_a (id, shape)
VALUES
    (1, 'Trójkąt'),
    (2, 'Kwadrat'),
    (3, 'Deltoid'),
    (4, 'Traper');
 
INSERT INTO shape_b (id, shape)
VALUES
    (1, 'Kwadrat'),
    (2, 'Trójkąt'),
    (3, 'Romb'),
    (4, 'Równoległobok');
```
Komenda INSERT INTO pozwala na dodanie do tabeli rekordów. Ogólna postać to:

```sql
INSERT INTO tab_name (col1_name, col2_name2, ...) 
VALUES
    (val1_col1, val2_col2),
    (val2_col1, val2_col2),
    ...
```

## Inner join 

Jest to podstawowy rodzaj złączenie. Ten sposób złączenia wybiera  te wiersze, dla których warunek złączenia jest spełniony. W żadnej z łączonych tabel kolumna użyta do łączenia nie może mieć wartości NULL. 

#### Przykład:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
INNER JOIN shape_b b ON a.shape = b.shape;
```
W zapytaniu powyżej użyto *aliasów* nazw tabel i column wynikowych, jest to szczególnie przydatne przy długich nazwach tabel i wprowadza czytelność w zapytaniu.

#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Trójkąt|2|Trójkąt|
|2|Kwadrat|1|Kwadrat|

## OUTER JOIN

Istnieją trzy rodzaje złączeń OUTER:
- LEFT OUTER JOIN,
- RIGHT OUTER JOIN,
- FULL OUTER JOIN.

### LEFT OUTER JOIN

Ten rodzaj złączenie zwróci wszystkie rekordy z lewej tablicy i dopasuje do nich rekordy z prawej tablicy które spełniją zadany warunek złączenia. Jeżeli w prawej tablicy nie występują rekordy spełnijące warunek złączenia z lewą w ich miejscu pojawią się wartości NULL.

#### Przykład 1:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
LEFT JOIN shape_b b ON a.shape = b.shape;
```
#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Trójkąt|2|Trójkąt|
|2|Kwadrat|1|Kwadrat|
|3|Deltoid|NULL|NULL|
|4|Traper|NULL|NULL|

#### Przykład 2:
```sql
SELECT
    b.id id_b,
    b.shape shape_b,
    a.id id_a,
    a.shape shape_a   
FROM
    shape_b b
LEFT JOIN shape_a a ON a.shape = b.shape;
```
#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Kwadrat|2|Kwadrat|
|2|Trójkąt|1|Trójkąt|
|3|Romb|NULL|NULL|
|4|Równoległobok|NULL|NULL|

### RIGHT OUTER JOIN

Działa jak left outer join z tym, że prawa tablica w zapytaniu jest brana w całości.

#### Przykład:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
RIGHT JOIN shape_b b ON a.shape = b.shape;
```

#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|2|Kwadrat|1|Kwadrat|
|1|Trójkąt|2|Trójkąt|
|NULL|NULL|3|Romb|
|NULL|NULL|4|Równoległobok|


### FULL OUTER JOIN

Jest złączeniem które zwraca:
- wiersze dla których warunek złączenia jest spełniony,
- wiersze z lewej tabeli dla których nie ma odpowiedników w prawej,
- wiersze z prawej tabeli dla których nie ma odpowiedników w lewej. 

#### Przykład:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
FULL JOIN shape_b b ON a.shape = b.shape;
```
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Trójkąt|2|Trójkąt|
|2|Kwadrat|1|Kwadrat|
|3|Deltoid"|NULL|NULL|
|4|Traper|NULL|NULL|
|NULL|NULL|3|Romb|
|NULL|NULL|4|Równoległobok|

## Podzapytania

Podzapytanie zagnieżdżone SELECT znajduje się wewnątrz zewnętrznego zapytania SELECT, np. po klauzuli WHERE, HAVING lub FROM. W przypadku tego rodzaju zapytań w pierwszej kolejności wykonywane są wewnętrzne zapytania SELECT, a ich wynik jest wykorzystywany do zewnętrznego zapytania SELECT. Stąd łatwo zuważyć, że mogą one służyć do poprawy wydajności obsługi zapytania. Należy dobierać podzapytania tak by najbardziej zagnieżdżone podzapytanie zawierało najmniejszy zbiór poszukiwań. 

#### Przykład:
Jeżeli chcemy znaleźć w bazie informację o tytułach filmów zwróconych w zadanym okresie możemy wykonać następujące zapytanie:
```sql
SELECT
   film_id,
   title
FROM
   film
WHERE
   film_id IN (
      SELECT
         inventory.film_id
      FROM
         rental
      INNER JOIN inventory ON inventory.inventory_id = rental.inventory_id
      WHERE
         return_date BETWEEN '2005-05-29'
      AND '2005-05-30'
   );
```

#### Wynik
|film_id|title|
|-|-|
|307|Fellowship Autumn|
|255|Driving Polish|
|388|Gunfight Moon|
|130|Celebrity Horn|
|563|Massacre Usual|
|397|Hanky October|
|...|...|

### Używanie podzapytań

Pod zapytania mogą być używane w :
- SELECT,
- UPDATE,
- DELETE,
- Funkcjach agregujących,
- Do definiowania tabel tymczasowych.

Używając podzapytań zapytania SQL szybko mogą stać się mało czytelne. Przez co będą trudne w zrozumieniu i późniejszym utrzymaniu. W celu analizy zapytań można użyć klauzuli __EXPLAIN__, która przeanalizuje zapytanie. Klauzula ta może służyć również do porównywania wydajności zapytań

#### Przykład:
```sql
EXPLAIN SELECT
   *
FROM
   film
```

## Funkcje agregujące

Funkcje agregujące wykonują obliczenia na zestawie wierszy i zwracają pojedynczy wiersz. PostgreSQL udostępnia wszystkie standardowe funkcje agregujące SQL w następujący sposób:
- AVG () - zwraca średnią wartość.
- COUNT () - zwraca liczbę wartości.
- MAX () - zwraca maksymalną wartość.
- MIN () - zwraca minimalną wartość.
- SUM () - zwraca sumę wszystkich lub różnych wartości.

Pełna lista funkcji agregującej: https://www.postgresql.org/docs/9.5/functions-aggregate.html

Często używamy funkcji agregujących z klauzulą GROUP BY w instrukcji SELECT. W tych przypadkach klauzula GROUP BY dzieli zestaw wyników na grupy wierszy i funkcja agregująca wykonuje obliczenia dla każdej grupy, np. maksimum, minimum, średnia itp. Funkcji agregujących można używać funkcji agregujących jako wyrażeń tylko w następujących klauzulach: SELECT i HAVING.

### GROUP BY
Klauzula GROUP BY dzieli wiersze zwrócone z instrukcji SELECT na grupy. Dla  każdej grupy można zastosować funkcję agregującą, np. SUM aby obliczyć sumę pozycji lub
COUNT aby uzyskać liczbę elementów w grupach.

Poniższa instrukcja ilustruje składnię klauzuli GROUP BY:
```sql
SELECT 
    column_1, 
    aggregate_function(column_2)
FROM 
    tbl_name
GROUP BY 
    column_1;
```
Klauzula GROUP BY musi pojawić się zaraz po klauzuli FROM lub WHERE, n0astępnie GROUP BY zawiera listę  kolumna oddzielonych przecinkami. 

### HAVING
Często używamy klauzuli HAVING w połączeniu z klauzulą GROUP BY do filtrowania wierszy grup
które nie spełniają określonego warunku.

Poniższa instrukcja ilustruje typową składnię klauzuli HAVING:
```sql
SELECT
    column_1,
    aggregate_function (column_2)
FROM
    tbl_name
GROUP BY
    column_1
HAVING
    condition;
```
Klauzula HAVING ustawia warunek dla wierszy grup utworzonych przez klauzulę GROUP BY.  

Klauzula GROUP BY ma zastosowanie, podczas gdy klauzula WHERE określa wcześniej warunki dla poszczególnych wierszy.

## Zadania wprowadzające
Wykonaj zapytania przy użyciu DBMS:  
  
1. Znajdź listę wszystkich filmów o tej samej długości.
2. Znajdź wszystkich klientów mieszkających w tym samym mieście.
3. Oblicz średni koszt wypożyczenia wszystkich filmów.
4. Oblicz i wyświetl liczbę filmów we wszystkich kategoriach.
5. Wyświetl liczbę wszystkich klientów pogrupowanych według kraju.
6. Wyświetl informacje o sklepie, który ma więcej niż 100 klientów i mniej niż 300 klientów.
7. Wybierz wszystkich klientów, którzy oglądali filmy ponad 200 godzin.
8. Oblicz średnią wartość wypożyczenia filmu.
9. Oblicz średnią wartość długości filmu we wszystkich kategoriach.
10. Znajdź najdłuższe tytuły filmowe we wszystkich kategoriach.
11. Znajdź najdłuższy film we wszystkich kategoriach. Porównaj wynik z pkt 10.

Połączenie się z bazą:

In [2]:
from sqlalchemy import create_engine
import pandas as pd

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"

db = create_engine(db_string)

connection_sqlalchemy = db.connect()

connection = create_engine("postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb")


Zapytanie nr. 1:

In [3]:
# 1. Znajdź listę wszystkich filmów o tej samej długości.
result_set = db.execute("\
    SELECT \
            length, \
            ARRAY_AGG(title) \
    FROM \
        film \
    GROUP BY \
        length \
    ORDER BY length ASC")
for r in result_set:
    print(r)

(46, ['Alien Center', 'Iron Moon', 'Kwai Homeward', 'Labyrinth League', 'Ridgemont Submarine'])
(47, ['Divorce Shining', 'Downhill Enough', 'Halloween Nuts', 'Hanover Galaxy', 'Hawk Chill', 'Shanghai Tycoon', 'Suspects Quills'])
(48, ['Ace Goldfinger', 'Heaven Freedom', 'Rush Goodfellas', 'Midsummer Groundhog', 'Notting Speakeasy', 'Odds Boogie', 'Paradise Sabrina', 'Pelican Comforts', 'Stepmom Dream', 'Sunset Racer', 'Valentine Vanishing'])
(49, ['Grosse Wonderful', 'Doors President', 'Heavenly Gun', 'Hook Chariots', 'Hurricane Affair'])
(50, ['Adaptation Holes', 'Blues Instinct', 'Crossing Divorce', 'Lion Uncut', 'Muppet Mile', 'Natural Stock', 'Pilot Hoosiers', 'Smoking Barbarella', 'Zorro Ark'])
(51, ['Champion Flatliners', 'Deep Crusade', 'English Bulworth', 'Excitement Eve', 'Frisco Forrest', 'Hall Cassidy', 'Simon North'])
(52, ['Caddyshack Jedi', 'Harper Dying', 'Lust Lock', 'Side Ark', 'Spartacus Cheaper', 'Trojan Tomorrow', 'Westward Seabiscuit'])
(53, ['Beneath Rush', 'Cabin

Zapytanie nr. 2:

In [4]:
# Znajdź wszystkich klientów mieszkających w tym samym mieście.
from main import client_from_city
# dla przykładowego miasta Athenai:
print(client_from_city('Athenai'))

      city first_name last_name
0  Athenai      Linda  Williams


Zapytanie nr. 3:

Jeśli koszt wypożyczenia to rental_rate - a tak zintepretowałem to pytanie - to odpowiedź poniżej. Jeśli koszt wypożyczenia to amount z payment, to w zapytaniu 8. rozwiązałem zapytanie 3., a zapytaniem 3. rozwiązałem zapytanie 8.

In [5]:
# 3. Oblicz średni koszt wypożyczenia wszystkich filmów:
df = pd.read_sql('\
            SELECT \
                AVG(f.rental_rate) AS avg \
            FROM \
                film f', con=connection)
df

avg
0  2.98

Średni koszt wypożyczenia wszystkich filmów to 2.98

Zapytanie nr. 4:

In [6]:
# 4. Oblicz i wyświetl liczbę filmów we wszystkich kategoriach.
df = pd.read_sql('\
            SELECT \
                c.name, \
                COUNT(f.film_id) \
            FROM \
                film f \
                JOIN language l ON f.language_id = l.language_id \
                JOIN film_category f_c ON f.film_id = f_c.film_id \
                JOIN category c ON f_c.category_id = c.category_id \
            GROUP BY \
                c.name \
            ORDER BY \
                c.name ASC', con=connection)
df.columns = ['category', 'amount_of_films']
df

category  amount_of_films
0        Action               64
1     Animation               66
2      Children               60
3      Classics               57
4        Comedy               58
5   Documentary               68
6         Drama               62
7        Family               69
8       Foreign               73
9         Games               61
10       Horror               56
11        Music               51
12          New               63
13       Sci-Fi               61
14       Sports               74
15       Travel               57

Zapytanie nr. 5:

In [31]:
# 5. Wyświetl liczbę wszystkich klientów pogrupowanych według kraju.
df = pd.read_sql('\
            SELECT \
                country, \
                COUNT(c.customer_id) \
            FROM \
                customer c \
                JOIN address a ON c.address_id = a.address_id \
                JOIN city ci ON a.city_id = ci.city_id \
                JOIN country co ON ci.country_id = co.country_id \
            GROUP BY country \
            ORDER BY country ASC', con=connection)
df.columns = ['country', 'amount_of_customers']
df

country  amount_of_customers
0             Afghanistan                    1
1                 Algeria                    3
2          American Samoa                    1
3                  Angola                    2
4                Anguilla                    1
..                    ...                  ...
103               Vietnam                    6
104  Virgin Islands, U.S.                    1
105                 Yemen                    4
106            Yugoslavia                    2
107                Zambia                    1

[108 rows x 2 columns]

Zapytanie nr. 6:

In [47]:
# 6. Wyświetl informacje o sklepie, który ma więcej niż 100 klientów i mniej niż 300 klientów.
df = pd.read_sql("\
    SELECT \
        s.* \
    FROM \
        store s \
    JOIN staff st ON s.store_id = st.store_id \
    JOIN rental r ON st.staff_id = r.staff_id \
    JOIN customer c ON r.customer_id = c.customer_id \
    GROUP BY \
        s.store_id \
    HAVING \
        COUNT(DISTINCT c.customer_id) BETWEEN '300' AND '300'", con=connection)
df

Empty DataFrame
Columns: [store_id, manager_staff_id, address_id, last_update]
Index: []

Zapytanie zwraca pusty dataframe, co znaczy, że nie ma takiego sklepu. Po późniejszym sprawdzeniu wyszło, że oba sklepy miały powyżej 300.

Zapytanie nr. 7:

In [9]:
# 7. Wybierz wszystkich klientów, którzy oglądali filmy ponad 200 godzin.
from main import client_by_sum_length
print(client_by_sum_length(200))

    first_name last_name   sum
0        Brian     Wyman  1265
1      Antonio      Meek  1451
2        Leona    Obrien  1588
3    Katherine    Rivera  1615
4      Tiffany    Jordan  1667
..         ...       ...   ...
594     Wesley      Bull  4808
595      Clara      Shaw  4808
596      Tammy   Sanders  5065
597    Eleanor      Hunt  5360
598       Karl      Seal  5388

[599 rows x 3 columns]


Wszyscy klienci ogladali filmy ponad 200 godzin - sprawdziłem, że wszystkich klientów jest 599.

Zapytanie nr. 8:

Jeśli średnia wartość wypożyczenia filmu to rental_rate, to zamieniłem zapytanie 8. z zapytaniem 3. tak jak powyżej opisałem.

In [10]:
# 8. Oblicz średnią wartość wypożyczenia filmu.
df = pd.read_sql("\
    SELECT \
        AVG(p.amount) \
    FROM \
        film f \
        JOIN inventory i ON f.film_id = i.film_id \
        JOIN rental r ON i.inventory_id = r.inventory_id \
        JOIN payment p ON r.rental_id = p.rental_id", con=connection)
df.columns = ['avg_amount']
df

avg_amount
0    4.200606

Zapytanie nr. 9:

In [11]:
# 9. Oblicz średnią wartość długości filmu we wszystkich kategoriach.
df = pd.read_sql('\
            SELECT \
                c.name, \
                AVG(f.length) \
            FROM \
                film f \
                JOIN language l ON f.language_id = l.language_id \
                JOIN film_category f_c ON f.film_id = f_c.film_id \
                JOIN category c ON f_c.category_id = c.category_id \
            GROUP BY \
                c.name \
            ORDER BY \
                c.name ASC', con=connection)
df.columns = ['category', 'avg_length']
df

category  avg_length
0        Action  111.609375
1     Animation  111.015152
2      Children  109.800000
3      Classics  111.666667
4        Comedy  115.827586
5   Documentary  108.750000
6         Drama  120.838710
7        Family  114.782609
8       Foreign  121.698630
9         Games  127.836066
10       Horror  112.482143
11        Music  113.647059
12          New  111.126984
13       Sci-Fi  108.196721
14       Sports  128.202703
15       Travel  113.315789

Zapytanie nr. 10:

In [38]:
# 10. Znajdź najdłuższe tytuły filmowe we wszystkich kategoriach:
df = pd.read_sql('\
            SELECT \
                name, \
                title \
            FROM \
                film f \
                JOIN film_category f_c ON f.film_id = f_c.film_id \
                JOIN category c ON f_c.category_id = c.category_id \
            WHERE \
                char_length(title) = (SELECT MAX(char_length(title)) FROM film \
                    JOIN film_category f_c2 ON film.film_id = f_c2.film_id \
                    JOIN category c2 ON f_c2.category_id = c2.category_id \
                    WHERE c2.name = c.name) \
            GROUP BY name, title', con=connection)
df.columns = ['category name', 'longest_title']
df

category name                longest_title
0         Action      Entrapment Satisfaction
1      Animation       Telemark Heartbreakers
2       Children         Heartbreakers Bright
3       Children         Microcosmos Paradise
4       Children         Sweethearts Suspects
5       Classics      Extraordinary Conquerer
6         Comedy      Trainspotting Strangers
7    Documentary       Deliverance Mulholland
8    Documentary       Intolerable Intentions
9          Drama       Goldfinger Sensibility
10        Family       Resurrection Silverado
11       Foreign         Ballroom Mockingbird
12       Foreign         Impossible Prejudice
13       Foreign         Reservoir Adaptation
14         Games        Bulworth Commandments
15         Games        Creatures Shakespeare
16        Horror  Arachnophobia Rollercoaster
17         Music       Confidential Interview
18           New        Flamingos Connecticut
19        Sci-Fi         Barbarella Streetcar
20        Sci-Fi         Cincinatti Whisperer
21        Sci-Fi         Silverado Goldfinger
22        Sci-Fi         Unforgiven Zoolander
23        Sci-Fi         Wonderland Christmas
24        Sports    Satisfaction Confidential
25        Travel      Desperate Trainspotting

Zapytanie nr 11.:

In [42]:
# 11. Znajdź najdłuższy film we wszystkich kategoriach. Porównaj wynik z pkt 10.
df = pd.read_sql('\
            SELECT \
                name, \
                title, \
                length \
            FROM \
                film f \
                JOIN film_category f_c ON f.film_id = f_c.film_id \
                JOIN category c ON f_c.category_id = c.category_id \
            WHERE \
                f.length = (SELECT MAX(length) FROM film \
                    JOIN film_category f_c2 ON film.film_id = f_c2.film_id \
                    JOIN category c2 ON f_c2.category_id = c2.category_id \
                    WHERE c2.name = c.name) \
            GROUP BY name, title, length', con=connection)
df.columns = ['category name', 'longest_title', 'length']
df

category name       longest_title  length
0         Action      Darn Forrester     185
1         Action        Worst Banger     185
2      Animation         Gangs Pride     185
3      Animation        Pond Seattle     185
4       Children         Fury Murder     178
5       Children      Wrong Behavior     178
6       Classics   Conspiracy Spirit     184
7         Comedy      Control Anthem     185
8    Documentary           Wife Turn     183
9    Documentary      Young Language     183
10         Drama       Jacket Frisco     181
11        Family      King Evolution     184
12       Foreign    Crystal Breaking     184
13       Foreign      Sorority Queen     184
14         Games       Chicago North     185
15        Horror    Analyze Hoosiers     181
16        Horror       Love Suicides     181
17         Music           Home Pity     185
18           New      Frontier Cabin     183
19        Sci-Fi  Soldiers Evolution     185
20        Sports     Smoochy Control     184
21        Travel       Muscle Bright     185
22        Travel   Sweet Brotherhood     185

W zapytaniu 11 i 10 wyniki są różne, ponieważ w 10 bierzemy pod uwagę długość tytułu, a w 11 długość filmu. W obu przypadkach wyświetlamy więcej niż jeden film, jeśli są tej samej długości.

## Zadanie implementacyjne
Zaimplementuj wszystkie funkcje w pliku main.py zgodnie z opisem a następnie przetestuj je w notatniku.

In [ ]:
import main
# wyświetla filmy w kategorii według podanej category_id:
print(main.film_in_category(3))

                   title               languge  category
0    Backlash Undefeated  English               Children
1         Bear Graceland  English               Children
2           Beneath Rush  English               Children
3          Betrayed Rear  English               Children
4            Cabin Flash  English               Children
5       Casper Dragonfly  English               Children
6    Christmas Moonshine  English               Children
7           Circus Youth  English               Children
8     Clockwork Paradise  English               Children
9      Comancheros Enemy  English               Children
10       Crooked Frogmen  English               Children
11      Daughter Madigan  English               Children
12          Doctor Grail  English               Children
13      Empire Malkovich  English               Children
14          Fargo Gandhi  English               Children
15     Forever Candidate  English               Children
16       Full Flatliners  Engli

In [ ]:
# wyświetla liczbę filmów w kategorii według podanej category_id:
print(main.number_films_in_category(4))

   category  count
0  Classics     57


In [ ]:
# wyświetla liczbe filmów w przedziale długości filmów:
print(main.number_film_by_length(60, 70))

    length  count
0       60      8
1       61     10
2       62      6
3       63      9
4       64      9
5       65      7
6       66      2
7       67      8
8       68      5
9       69      6
10      70      7


In [ ]:
# wyświetla klientów z wpisanego miasta
print(main.client_from_city('Athenai'))

      city first_name last_name
0  Athenai      Linda  Williams


In [ ]:
# wyświetla średni koszt wypożyczenia dla filmu o wskazanej długości:
print(main.avg_amount_by_length(60))


   length       avg
0      60  4.089099


In [ ]:
# wyświetla klientów którzy wypożyczyli filmy z sumaryczną długością dłuższą niż podana długość
print(main.client_by_sum_length(5000))

  first_name last_name   sum
0      Tammy   Sanders  5065
1    Eleanor      Hunt  5360
2       Karl      Seal  5388


In [ ]:
# wyświetla różne statystyki (avg, sum, min, max) na temat filmów według ich kategorii
print(main.category_statistic_length('Travel'))

  category         avg   sum  min  max
0   Travel  113.315789  6459   47  185
